src/Microsoft.AspNet.Mvc.DataAnnotations/CompareAttributeAdapter.cs

```diff
@@ -11,26 +10,29 @@ namespace Microsoft.AspNet.Mvc.ModelBinding.Validation
 {
     public class CompareAttributeAdapter : AttributeAdapterBase<CompareAttribute>
     {
+        private readonly string _otherProperty;
+
         public CompareAttributeAdapter(CompareAttribute attribute, IStringLocalizer stringLocalizer)
             : base(new CompareAttributeWrapper(attribute), stringLocalizer)
         {
-            if (attribute == null)
-            {
-                throw new ArgumentNullException(nameof(attribute));
-            }
+            _otherProperty = "*." + Attribute.OtherProperty;
```

Comment:
`OtherProperty` does not always have the correct value when this constructor executes. That's why `CompareAttributeWrapper` exists.